### ISSUE TITLE
Feature Request : Events can be updated for each month

### ISSUE THREAD

**Issue by pvsvamsi** [aleksanderwozniak/table_calendar#86]
Thanks for such an awesome library.
Can we have provision to update the events map, based on the month selected.
Typically on load of each month, we send a request to server to get the events for that month.
I tried updating the _events on each request's result, but the event markers are not reflecting as the controller is still referring to the previous events map.

**Comment by PeterOmbodi** [comment]
After adding using CalendarController to the library, I encountered the same problem. Judging by the code, _events for the controller is set only at initialization

**Comment by pvsvamsi** [comment]
Yes. But that restricts a Major potential use case for this awesome library.

**Comment by fabianogadenz** [comment]
I'm going through the same problem. I'm using onVisibleDaysChanged, getting the start date to make a request. The problem occurs when the outsideDaysVisible option is true, as my calendar may be 3 months old. But leaving outsideDaysVisible = false, the implementation on the onVisibleDaysChanged method worked perfectly.

**Comment by MarcosMota** [comment]
i'am with same issue

**Comment by MarcosMota** [comment]
Am I having the same problem even implementing outsideDaysVisible = false?

**Comment by PeterOmbodi** [comment]
As a temporary solution, until the library has will be fixed by honored @aleksanderwozniak, I added setter for _events in calendar_controler.dart
void setEvents(Map<DateTime, List> events){
    _events = events;
  }
and in my widget state class
@override
  void didUpdateWidget(WorkloadCalendar oldWidget) {
    //todo temporary solution until fix library
    if (oldWidget != null && _selectedDay!=null 
      _calendarController.setSelectedDay(_selectedDay);
}
....
Widget _buildTableCalendar() {
    //todo temporary solution until fix library
    if (_tableCalendar != null ??
        _tableCalendar.events != _calendarController.visibleEvents)
      _calendarController.setEvents(_visibleEvents);

    _tableCalendar = TableCalendar(
      calendarController: _calendarController,
      events: _visibleEvents,
...
)}
Now, showing asynchronous receiving events works correctly.

**Comment by MattGson** [comment]
+1 on this

**Comment by ZantsuRocks** [comment]
+2 on this.

**Comment by ZantsuRocks** [comment]
> 
> 
> I'm going through the same problem. I'm using onVisibleDaysChanged, getting the start date to make a request. The problem occurs when outsideDaysVisible = true, as my calendar may display 3 months on screen. But leaving outsideDaysVisible = false, the implementation on the onVisibleDaysChanged method worked perfectly. Does anyone know how I can get the selected month?

You can add 7 days on the first day (first argument), this make your DateTime ALWAYS be on the correct month.

**Comment by MattGson** [comment]
I'd like to add that this shouldn't be specific to changing views. The calendar should be able to be updated whenever new data is loaded i.e. when async data load completes. @PeterOmbodi  solution works well in this regard.

**Comment by ZantsuRocks** [comment]
I finally have done this.
You can update the events if you DO NOT reinstantiate the event Map.
If you need to remove all the events use the method .clear() instead of assign the {} to the object.

**Comment by pvsvamsi** [comment]
I have reverted back to the older version (table_calendar: ^1.2.5) and it fixed the issue for me.

**Comment by MattGson** [comment]
> I finally have done this.
> You can update the events if you DO NOT reinstantiate the event Map.
> If you need to remove all the events use the method .clear() instead of assign the {} to the object.

Cheers @ZantsuRocks . This works perfectly.

**Comment by pmagnuson** [comment]
@ZantsuRocks I am unclear about your solution.  I have used .clear() and indeed the event map is clear and events are no longer shown.  How are you getting new events to display?  Are you using the methos of @PeterOmbodi ?

**Comment by MattGson** [comment]
@pmagnuson You need to keep a reference to the events list and call clear(), then add the new events to the list. Use setState if you want the calendar to rebuild. No need for anything else.

**Comment by pmagnuson** [comment]
@MattGson . thank you.  I finally got this working by using
`
events.clear();
events.addAll( newEventMap );
`

Thanks for all the help.

**Comment by aleksanderwozniak** [comment]
Can you try code from the [dev branch](https://github.com/aleksanderwozniak/table_calendar/tree/dev)?

It should solve this issue -> use `_calendarController.events` property to update the data.

**Comment by MarceloRab** [comment]
I can't update events via firebase stream. When you update events the calendar keeps the old event visible.

**Comment by aleksanderwozniak** [comment]
@MarceloRab 
Did you use the dev branch? Can you provide your code?

**Comment by MarceloRab** [comment]
https://pastebin.com/PTwWYaD0

There are streams to watch for changing events in firebase, but it does not redo events when updated.

**Comment by MarceloRab** [comment]
It does not update in real time firebase.

**Comment by aleksanderwozniak** [comment]
@MarceloRab 
I see that you are not using the `dev` branch. It does have the solution to this issue, but I want to properly test it before pushing the next update.

Add this to your `pubspec.yaml` file:
```yaml
table_calendar:
  git:
    url: git://github.com/aleksanderwozniak/table_calendar.git
    ref: dev
```

Then use this in your code:
```dart
setState(() {
  _calendarController.events = updatedEvents;
});
```

`setState` might be redundant. You just need to rebuild the **TableCalendar** widget.

**Comment by MarceloRab** [comment]
Sorry, I could not enter the dev.

table_calendar: ^2.0.2
    git:
      url: git: //github.com/aleksanderwozniak/table_calendar.git
      ref: dev

**Comment by aleksanderwozniak** [comment]
@MarceloRab 
Simply copy the pubspec lines that I wrote -> that is use the second version. Note that you have a whitespace between 'git:' and '//github[...]', which should not be there.

**Comment by MarceloRab** [comment]
Thank you very much.

I added

  _calendarController
               ? .setSelectedDay (_day.chosen,
                   runCallback: true,
                   animate: false);

To clear events in realTime when the day is selected.

**Comment by aleksanderwozniak** [comment]
TableCalendar `2.1.0` solves this issue. Simply update `events` property and rebuild the widget.

**Comment by GonzaloSaad** [comment]
> TableCalendar `2.1.0` solves this issue. Simply update `events` property and rebuild the widget.

Is it possible that this update does not have the property?

**Comment by aleksanderwozniak** [comment]
@GonzaloSaad 
It should have it. Check if `pubspec.lock` got updated to the proper value.

**Comment by GonzaloSaad** [comment]
Yes, I've got the latest. But that property is not in place. 

Thanks for answering!

**Comment by MarceloRab** [comment]
How are you. I'm doing a new project on Flutter and can't get the setter event package. An error in "ref: dev"

`table_calendar:
    git:
      url: git://github.com/aleksanderwozniak/table_calendar.git
      ref: dev`

**Comment by GonzaloSaad** [comment]
I beileve the `dev` branch is no longer available, and version `2.1.0` does not have that setter. 

Can you plase confirm @MarceloRab , that the setter is no there?

**Comment by aleksanderwozniak** [comment]
@GonzaloSaad 
Ah now I get what you meant. You can update `events` property of TableCalendar **widget** (that is rebuild the widget with new data passed to `events`), not the CalendarController.

For full example refer here: https://gist.github.com/aleksanderwozniak/1367892cbe39c25afda403b2300652ca

Note that you don't need to use this BLoC-like pattern, this was just to showcase using TableCalendar with Streams.

@MarceloRab 
The `dev` branch was there only for pre-release testing, you can specify the dependency in standard way now: 
```yaml
table_calendar: ^2.2.0
```

**Comment by GonzaloSaad** [comment]
Ahhhh, understood. 

Thank you very much then. Sorry!
